refactor: frappe.db.set_value

gavindsouza opened the pull request:
### Changes
* Simplified logic
* Perf enhancements (removed unnecessary conditional computations) for basic and Single DocTypes
* Use query builder and ORM to build queries instead of juggling
* Expanded scope of parameterized query building via `frappe.qb`
* Fixed bug in `frappe.utils.data.sbool` converting 'on', 'off', 't', 'f', 'yes', 'no' to True/False. Now only "true", "false", "1", "0" will be handled
* Fixed de-serializing of timedelta objects (new util: `format_timedelta`) 
* Fixed bug in `frappe.db.get_single_value` returning 0 for True values, caching
* Added DB API `frappe.db.set_single_value` to escape use of `frappe.db.set_value("Single DocType", "Single DocType",...)` and `frappe.db.set_value("Single DocType", None,...)` patterns
* Addition of 9 tests for `frappe.db.set_value`
* Addition of 3 tests for `frappe.qb` parameterization (by @saxenabhishek)
* Addition of 3 tests for new util `format_delta`

|No. of Queries|Before|After|
|---|---|---|
|Minimum|N + 1|1|
|Maximum|N + 1|2|

* Yes, this PR changes `frappe.db.set_value`'s time complexity from O(N) to O(1) 
* Perf stats in comment below at https://github.com/frappe/frappe/pull/15560#issuecomment-1010024291

### Discussion

#### Number of UPDATE queries triggered

At this point, the number of queries `set_value` triggers depends on how many rows it is going to touch (and an extra one). If your set_value query is meant to update the whole table, and your table has N rows (documents), set_value will trigger N+1 queries. 1 query that takes a FOR UPDATE lock and the other N to update the data.

The N individual UPDATE queries are not required. I'll attempt to bring them down to 1.

#### What about the extra SELECT query

Now, by default for_update is set to True. But in case you feel you don't need to take that lock, and need just a plain old UPDATE query, then you don't need the SELECT. This `SELECT` (through `self.get_values`) can be avoided if `for_update=False` is set, and set_value would then just trigger a single UPDATE query (when the previous block's goals are met).

Optional 1 SELECT query.

#### Reference

https://github.com/frappe/frappe/blob/b64cab6867dfd860f10ccaf41a4ec04bc890b583/frappe/database/database.py#L700-L720

---

Re of https://github.com/frappe/frappe/pull/14060

gavindsouza committed:
refactor: frappe.db.set_value

* Simplified logic
* Perf enhancements (removed unnecessary conditional computations)
* Use query builder and ORM to build queries instead of juggling

gavindsouza committed:
fix: Expand iterable to get rid of extra brackets

Although valid SQL, MariaDB didn't like double brackets on VALUES. It
raised ERROR 1241: Operand should contain 1 column(s)

gavindsouza committed:
fix: Run query even if no documents found for_update

If no docnames are found, then use NULL as fallback. I tested this on PostgreSQL & MariaDB and it seems to work as expected

    In [1]: from frappe.query_builder import Field

    In [2]: frappe.db.set_value('ToDo', Field("creation") > Field("modified"), 'description', 'change 2', for_update=True, debug=True)
    SELECT "name" FROM "tabToDo" WHERE "creation">"modified"
    Execution time: 0.0 sec
    UPDATE "tabToDo" SET "description"='change 2',"modified"='2021-10-22 09:38:40.110481',"modified_by"='Administrator' WHERE "name" IN (NULL)
    Execution time: 0.0 sec

gavindsouza committed:
fix: Cast values as str for all single doctypes

Re-arranged block for simplicity. Type casting doesn't change anything
it seems. On get_value, values are casted properly.

gavindsouza committed:
fix: Clear cache on set_value

gavindsouza committed:
fix(typing): Add type hints for frappe.cache

ankush commented on frappe/__init__.py:
```diff
@@ -310,7 +313,7 @@ def destroy():
 
 # memcache
```
```suggestion
```
🤔

gavindsouza committed:
fix(qb): Patch ParameterizedValueWrapper as wrapper_cls

gavindsouza committed:
style: Sorted imports & whitespace consistency

gavindsouza committed:
style: remove commented code, whitespaces test_file, added typing hint

gavindsouza committed:
fix(qb:ValueWrapper): Use get_value_sql only for str values

gavindsouza committed:
chore(typing): Add type hints in qb builder classes

gavindsouza committed:
fix: Don't cast to str if None or Falsy

* refactor: use get_single_value instead of set_value(blah, None, blah1)
* User.reload in test_home_page to combat what happens locally

gavindsouza commented:
Waiting for @saxenabhishek to add tests and docstrings for query builder parameterization related stuff

saxenabhishek committed:
style: test_query_builder

gavindsouza committed:
fix(single:set_value): Cast using sbool instead of str for all

gavindsouza committed:
test: Add tests for frappe.db.set_value

gavindsouza committed:
feat: frappe.db.set_single_value

Alias to frappe.db.set_value without having to pass the doctype name
twice or teh second one as None

Other changes:
* set/get db APIs for single DocTypes
* Make use of redundant cache key in frappe.db.get_single_value

saxenabhishek committed:
test: qb parameterization

where conditions
set update conditions
where with a function

saxenabhishek committed:
refactor: pythonic NamedParameterWrapper

saxenabhishek committed:
refactor: remove reduandant func call

saxenabhishek committed:
docs: Working of NamedParameters in qb

gavindsouza committed:
test: Postgres compatible db tests

gavindsouza committed:
fix: sbool before cint in cast for Check types

Noticed an issue when get_single_value wasn't returning the correct
values; by converting true to 0.

Tested it out. Here's the examples:

In [2]: sbool("2")
Out[2]: '2'

In [3]: cint(sbool("2"))
Out[3]: 2

In [4]: cint(sbool("-1"))
Out[4]: -1

In [5]: cint(sbool("0"))
Out[5]: 0

In [6]: cint(sbool("1000"))
Out[6]: 1000

In [7]: cint(sbool("10_000"))
Out[7]: 10000

In [8]: cint(sbool("true"))
Out[8]: 1

gavindsouza committed:
Merge branch 'set_value-refactor' of github.com:gavindsouza/frappe into set_value-refactor

saxenabhishek committed:
fix: replace the field method with a column

gavindsouza committed:
fix: Set cache=True to maintain old behaviour consistency

saxenabhishek committed:
fix: ignore copy of getattr methods

gavindsouza commented:
**Quicf Stats:**

New:
<img width="716" alt="new" src="https://user-images.githubusercontent.com/36654812/148964223-af22981b-44dc-472d-9ebf-8ce51096d32a.png">

Old:
<img width="716" alt="old" src="https://user-images.githubusercontent.com/36654812/148964229-ceb1a331-01f5-4698-8fa3-87383e1a3aaf.png">

<details><summary>A bit more detailed runs:</summary>
<p>
Ran a couple of set_value queries 10k times with the specified (kw)args. This test has been done for updating multiple columns & multiple rows. The refactored solution uses the new PyPika parameterized query engine, while the develop branch uses string substitution.

**Current:**

This would mostly end up in failures due to `frappe.db.MAX_WRITES_PER_TRANSACTION`, given the unknown number of queries set_value would trigger 🥲  

20k queries like `frappe.db.set_value("ToDo", {"description": ("like", "%test%")}, {"status": "Closed"})`
<img width="608" alt="develop-set_value" src="https://user-images.githubusercontent.com/36654812/148956620-eb952844-1a5f-4fc7-8db7-909d8352dedd.png">

20k queries like `frappe.db.set_value("ToDo", {"description": ("like", "%test%")}, {"status": "Closed"}, for_update=False)`
<img width="608" alt="develop-set_value-no_lock" src="https://user-images.githubusercontent.com/36654812/148956632-3e4849a9-71be-469b-81ea-dafecf75b344.png">

20k queries like `frappe.db.set_value("ToDo", {"description": ("like", "%test%")}, {"status": "Closed"}, for_update=False, update_modified=False)`
<img width="608" alt="develop-set_value-no_lock-no_modified" src="https://user-images.githubusercontent.com/36654812/148958728-ddfb0be4-2756-43d6-bd9d-b46379fbb025.png">


**Refactored:**

20k queries like `frappe.db.set_value("ToDo", {"description": ("like", "%test%")}, {"status": "Closed"})`
<img width="608" alt="refactor-set_value" src="https://user-images.githubusercontent.com/36654812/148956975-9b9dfa4f-daf7-4883-a1ca-007b620aaecd.png">

20k queries like `frappe.db.set_value("ToDo", {"description": ("like", "%test%")}, {"status": "Closed"}, for_update=False)`
<img width="608" alt="refactor-set_value-no_lock" src="https://user-images.githubusercontent.com/36654812/148956985-cd38fce9-5761-420f-8d26-1458ec9c1e45.png">

20k queries like `frappe.db.set_value("ToDo", {"description": ("like", "%test%")}, {"status": "Closed"}, for_update=False, update_modified=False)`

<img width="608" alt="refactor-set_value-no_lock-no_modified" src="https://user-images.githubusercontent.com/36654812/148957902-6353b658-dc48-4475-b5e5-aef4eaabce29.png">

---

**Script:**

```python
import frappe
import unittest
import time

class TestPerf(unittest.TestCase):
    def setUp(self):
        self.startTime = time.time()

    def tearDown(self):
        t = time.time() - self.startTime
        print('%s: %.3f' % (self.id(), t))

    def test_set_value_perf(self):
        # frappe.db.MAX_WRITES_PER_TRANSACTION = 100_00_000 - this is required for the `develop`
        for _ in range(10_000):
            frappe.db.set_value("ToDo", {"description": ("like", "%test%")}, {"status": "Closed"}, for_update=False, update_modified=False)
            frappe.db.set_value("ToDo", {"description": ("like", "%test%")}, {"status": "Open"}, for_update=False, update_modified=False)
```
</p>
</details>

ankush commented:
Few tests failing on ERPNext: https://github.com/frappe/erpnext/runs/4802127202?check_suite_focus=true 

1. incorrect escaping of QB objects (generated query, check line 3)
![telegram-cloud-photo-size-5-6269553406708395649-y](https://user-images.githubusercontent.com/9079960/149332529-58392139-5af2-4716-861d-edd5c43e9077.jpg)

2-3 more failing not 100% sure if they are related to this PR.

gavindsouza commented:
Looks like this may be related to the changes made in query parameterization. Should've moved it in a separate PR when I discovered the issues. 🥲

saxenabhishek committed:
fix: make parameters for strings only

ankush commented on frappe/database/database.py:
```diff
@@ -677,53 +692,55 @@ def set_value(self, dt, dn, field, val=None, modified=None, modified_by=None,
+				filters={"field": ("in", tuple(to_update)), "doctype": dt}, debug=debug
+			)
+
+			singles_data = ((dt, key, sbool(value)) for key, value in to_update.items())
```
Any specific reason for `sbool`? `sbool` isn't JUST converting truthy/falsy values it will also assume more:

<img width="1265" alt="Screenshot 2022-01-17 at 12 48 14 PM" src="https://user-images.githubusercontent.com/9079960/149724528-a86ae303-082f-4e92-9575-5fd763c0a8e5.png">

There are some singles where brilliant minds have used "yes"/"no" instead of checkbox which could break after this change 😄

gavindsouza replied:
Fixed (in sbool), only converting values like "true" and "1" now to bool

gavindsouza committed:
Merge branch 'develop' of github.com:frappe/frappe into set_value-refactor

gavindsouza committed:
fix!: Don't use strtobool in sbool util

strtobool would convert 't', 'yes' to True too. Which shouldn't be a
problem generally but could be one possibly.

gavindsouza committed:
style: Add typing, sorted imports#

gavindsouza commented:
Highly unlikely that `sbool` changes did this. I'm guessing CI failures are related to the parameterization changes (via https://github.com/frappe/frappe/pull/15560/commits/02bdc35490fff1e2b0c31c1ac08bcf0151873807), can't say for sure atm but everything was passing post that. cc: @saxenabhishek

saxenabhishek committed:
fix: sbool converting int stored as string

saxenabhishek committed:
fix: timedelta parsing in pypika

gavindsouza commented on frappe/query_builder/terms.py:
```diff
@@ -51,6 +52,9 @@ def get_sql(
 		else:
+			# * BUG: pypika doesen't parse timedeltas
+			if isinstance(self.value, timedelta):
+				self.value = str(self.value)
```
Not really a good idea. hour ranges out of the `0..23` range get stringified to something like `"5 days, 32:02.1211"`

saxenabhishek replied:
There is a nice ISO format for time durations ([ISO8601#Durations](https://en.wikipedia.org/wiki/ISO_8601#Durations)) but `datetime` in python doesn't support it for `timedelta` yet. We can add utils to translate it ourselves or maybe use this package [isodate](https://pypi.org/project/isodate/)

gavindsouza replied:
Added a format_timedelta util to handle this "better"; Added tests too.

saxenabhishek committed:
test: avoid parameterization of qb objects

gavindsouza commented:
**Pending:**
- [x] Standardize timedelta string formatting

gavindsouza committed:
feat(minor): Case option in run-tests for specifying TestCase

gavindsouza committed:
fix: Handle parsing and formatting timedeltas

* Added utils parse_timedelta, format_timedelta
* Added to json_handler for de-serializing timedelta objects

gavindsouza committed:
test: Added unit tests for format_timedelta, parse_timedelta,
json_handler

gavindsouza committed:
style: Sort imports

gavindsouza committed:
Merge branch 'set_value-refactor' into set_value-refactor

gavindsouza committed:
Merge branch 'develop' of github.com:frappe/frappe into set_value-refactor

gavindsouza commented:
<img width="1552" alt="Screenshot 2022-01-28 at 6 53 16 PM" src="https://user-images.githubusercontent.com/36654812/151554255-d26a8a5b-adf1-4559-80b1-e4c6b94b7bad.png">

Now, what does the failing codecov check mean `(53.69% of diff hit (target 85.00%))`? When the net coverage still seems to be increasing by 0.01% and server coverage by 0.02% (after adding 13 tests and adding more asserts in 2 tests) 🥲

gavindsouza committed:
Merge branch 'develop' of github.com:frappe/frappe into set_value-refactor

gavindsouza committed:
fix: Format timedelta object accurately

gavindsouza commented:
@Mergifyio backport version-13-hotfix
